File: src/main/java/ukalus/container/creators/ConstructorCreator.java
package ukalus.container.creators;

import ukalus.container.InstanceCreator;
import ukalus.container.Resolver;

import java.lang.reflect.Constructor;
import java.lang.reflect.InvocationTargetException;
import java.util.stream.Stream;

public class ConstructorCreator implements InstanceCreator {

  private Class type;
  private Resolver[] arguments;

  public ConstructorCreator(Class type, Resolver[] arguments) {
    this.type = type;
    this.arguments = arguments;
    if (arguments == null) {
      throw new NullPointerException();
    }
  }

  public Object newInstance() throws IllegalArgumentException,
      InstantiationException, IllegalAccessException, InvocationTargetException {
    Constructor constructor = Stream.of(type.getConstructors()).filter(new ConstructorMatches(arguments.length)).findFirst().orElse(null);
    Class<?>[] types = constructor.getParameterTypes();

    Object[] values = new Object[arguments.length];
    for (int i = 0; i < arguments.length; i++) {
      values[i] = arguments[i].getValue(types[i]);
    }
    
    return constructor.newInstance(values);
  }

}

File: src/main/kotlin/ukalus/container/creators/ConstructorCreator.kt
package ukalus.container.creators

import ukalus.container.InstanceCreator
import ukalus.container.Resolver

class ConstructorCreator(private val type: Class<*>, private val arguments: Array<Resolver>) : InstanceCreator {
    override fun newInstance(): Any {
        val constructor = type.constructors.find { it.parameterCount == arguments.size }!!
        val types = constructor.parameterTypes
        val values = arguments.mapIndexed { index, resolver -> resolver.getValue(types[index]) }
        return constructor.newInstance(*values.toTypedArray())
    }
}
